modules/mod-deep-learning/ModelCard.h

```diff
@@ -0,0 +1,212 @@
+
+   std::string author() const { return m_author; }
+   void author(const std::string &author) { m_author = author; }
+
+   std::string long_description() const { return m_long_description; }
+   void long_description(const std::string &long_description) { m_long_description = long_description; }
+
+   std::string short_description() const { return m_short_description; }
+   void short_description(const std::string &short_description) { m_short_description = short_description; }
+
+   int sample_rate() const { return m_sample_rate; }
+   void sample_rate(int rate) { m_sample_rate = rate; }
+
+   bool multichannel() const { return m_multichannel; }
+   void multichannel(bool multichannel) { m_multichannel = multichannel; }
+
+   std::string effect_type() const { return m_effect_type; }
+   void effect_type(const std::string &type) { m_effect_type = type; }
+
+   const std::vector<std::string> domain_tags() const { return m_domain_tags; }
+   void domain_tags(const std::vector<std::string> tags) { m_domain_tags = tags; }
+
+   const std::vector<std::string> tags() const { return m_tags; }
+   void tags(const std::vector<std::string> tags) { m_tags = tags; }
+
+   const std::vector<std::string> labels() const { return m_labels; }
+   void labels(std::vector<std::string> labels) { m_labels = labels; } // use std::move to not copy
+
+   size_t model_size() const { return m_model_size; }
+   void model_size(size_t model_size) { m_model_size = model_size; }
+
+private:
+   std::string m_name {""};
+   std::string m_author {""};
+   std::string m_long_description {""};
+   std::string m_short_description {""};
+   int m_sample_rate {0};
+   bool m_multichannel  {false};
+   std::string m_effect_type {""};
+   std::vector<std::string> m_domain_tags;
+   std::vector<std::string> m_tags;
+   std::vector<std::string> m_labels;
+   size_t m_model_size {0};
+
+   bool m_is_local {false};
+   std::string m_local_path {""};
+
+public:
+
+   // TODO: add versioning?
+   bool operator==(const ModelCard& that) const;
+   bool operator!=(const ModelCard& that) const;
+};
+
+using ModelCardHolder = std::shared_ptr<ModelCard>;
+using ModelCardFilter = std::function<bool(ModelCardHolder card)>;
+
+class ModelCardCollection final
+{
+public:
+   // set the JSON schema to be used when Validate() is called. 
+   ModelCardCollection() = default;
+
+   // insert a model card into the collection if it matches the schema
+   // will throw if the card does not match the schema
+   // only adds if the card isn't already there (doesn't throw)
+   // @execsafety strong
+   void Insert(ModelCardHolder card);
+
+   // returns a view of a subset as dictated by the filter
+   ModelCardCollection Filter(ModelCardFilter filter) const;
+
+   // returns an iterator to the cards
+   std::vector<ModelCardHolder>::const_iterator begin() const {return mCards.begin();}
+   std::vector<ModelCardHolder>::const_iterator end() const {return mCards.end();}
```

Comment:
Iteration over a const container (so you can't insert into it) but exposing shared pointers to non-const cards.  Do you want that, or do you want "deep const?"

Reply:
I don't think const is really needed here, so I removed it for now.